From e952ad8d58c1cd8c61012bbe30e18712ac7928e7 Mon Sep 17 00:00:00 2001
From: Jammy Huang <jammy_huang@aspeedtech.com>
Date: Thu, 16 Dec 2021 10:49:54 +0800
Subject: [PATCH 18/27] media: aspeed: Correct values for detected timing

Correct timing's fp/sync/bp value based on the information below.
It should be noticed that the calculation formula should be changed
per sync polarity.

The sequence of signal: sync - backporch - video data - frontporch

The following registers start counting from sync's rising edge:
1. VR090: frame edge's left and right
2. VR094: frame edge's top and bottom
3. VR09C: counting from sync's rising edge to falling edge

            +--+     +-------------------+     +--+
            |  |     |    v i d e o      |     |  |
         +--+  +-----+                   +-----+  +---+

        sync+--+
    left/top+--------+
right/bottom+----------------------------+

                  +-------------------+
                  |    v i d e o      |
      +--+  +-----+                   +-----+  +---+
         |  |                               |  |
         +--+                               +--+
        sync+-------------------------------+
    left/top+-----+
right/bottom+-------------------------+

Change-Id: Ib87370b36773481dc7bd2ea78454a4b87e35be11
Signed-off-by: Jammy Huang <jammy_huang@aspeedtech.com>
---
 drivers/media/platform/aspeed-video.c | 27 ++++++++++++++++++++++-----
 1 file changed, 22 insertions(+), 5 deletions(-)

diff --git a/drivers/media/platform/aspeed-video.c b/drivers/media/platform/aspeed-video.c
index 1c43f57258c8..dd3ba02c6910 100644
--- a/drivers/media/platform/aspeed-video.c
+++ b/drivers/media/platform/aspeed-video.c
@@ -1009,10 +1009,20 @@ static void aspeed_video_get_resolution(struct aspeed_video *video)
 						src_tb_edge);
 		video->frame_top = FIELD_GET(VE_SRC_TB_EDGE_DET_TOP,
 					     src_tb_edge);
-		det->vfrontporch = video->frame_top;
-		det->vbackporch = FIELD_GET(VE_MODE_DETECT_V_LINES, mds) -
-			video->frame_bottom;
 		det->vsync = FIELD_GET(VE_SYNC_STATUS_VSYNC, sync);
+		if (det->polarities & V4L2_DV_VSYNC_POS_POL) {
+			det->vbackporch = video->frame_top - det->vsync;
+			det->vfrontporch =
+				FIELD_GET(VE_MODE_DETECT_V_LINES, mds) -
+				video->frame_bottom;
+		} else {
+			det->vsync = FIELD_GET(VE_MODE_DETECT_V_LINES, mds) -
+				     det->vsync;
+			det->vbackporch = video->frame_top;
+			det->vfrontporch =
+				FIELD_GET(VE_MODE_DETECT_V_LINES, mds) -
+				video->frame_bottom - det->vsync;
+		}
 		if (video->frame_top > video->frame_bottom)
 			continue;
 
@@ -1020,9 +1030,16 @@ static void aspeed_video_get_resolution(struct aspeed_video *video)
 					       src_lr_edge);
 		video->frame_left = FIELD_GET(VE_SRC_LR_EDGE_DET_LEFT,
 					      src_lr_edge);
-		det->hfrontporch = video->frame_left;
-		det->hbackporch = htotal - video->frame_right;
 		det->hsync = FIELD_GET(VE_SYNC_STATUS_HSYNC, sync);
+		if (det->polarities & V4L2_DV_HSYNC_POS_POL) {
+			det->hbackporch = video->frame_left - det->hsync;
+			det->hfrontporch = htotal - video->frame_right;
+		} else {
+			det->hsync = htotal - det->hsync;
+			det->hbackporch = video->frame_left;
+			det->hfrontporch = htotal - video->frame_right -
+					   det->hsync;
+		}
 		if (video->frame_left > video->frame_right)
 			continue;
 
-- 
2.17.1

